### PR TITLE
Events for federated ingress controller

### PR DESCRIPTION
**Pull request by mwielgus**
cc: @quinton-hoole @kubernetes/sig-cluster-federation

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31746)

### CODE REVIEW

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 1]
```diff
@@ -287,6 +297,9 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
 			// We can't supply server-created fields when creating a new object.
 			desiredIngress.ObjectMeta.ResourceVersion = ""
 			desiredIngress.ObjectMeta.UID = ""
+			ic.eventRecorder.Eventf(baseIngress, api.EventTypeNormal, "CreateInCluster",
+				"Creating ingress in cluster %s", cluster.Name)
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Include the ingress namespace and name?  util.NamespacedName.String() can be used to ensure consistent formatting.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 2]
```diff
@@ -332,6 +345,9 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
 				for key, val := range baseIngress.ObjectMeta.Annotations {
 					desiredIngress.ObjectMeta.Annotations[key] = val
 				}
+				ic.eventRecorder.Eventf(baseIngress, api.EventTypeNormal, "UpdateCluster",
+					"Updating ingress in cluster %s", cluster.Name)
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: As above.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 3]
```diff
@@ -346,7 +362,10 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
 		return
 	}
 	glog.V(4).Infof("Calling federatedUpdater.Update() - operations: %v", operations)
-	err = ic.federatedUpdater.Update(operations, ic.updateTimeout)
+	err = ic.federatedUpdater.UpdateWithOnError(operations, ic.updateTimeout, func(op util.FederatedOperation, operror error) {
+		ic.eventRecorder.Eventf(baseIngress, api.EventTypeNormal, "FailedClusterUpdate",
+			"Update ingress in cluster %s failed: %v", op.ClusterName, operror)
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: As above.

**Comment by quinton-hoole** [comment]
Thanks @mwielgus .  Nits then LGTM.  Feel free to add the label yourself once addressed.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 1 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: The event is attached to the ingress object and all of the stuff is there. We don't include event owner in the event text.
Example:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go#L479

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 2 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: As above :)

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 3 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: As above :)

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 1 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Aah - makes sense.  As you can tell, I've never used events :-)

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 2 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: As above :-)

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 3 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: As above :-)

**Pull request merged**